mne/time_frequency/csd.py

```diff
@@ -1,63 +1,521 @@
+    mat = mat + np.rollaxis(mat, 1).conj()
+
+    # We counted the diagonal twice
+    if np.issubdtype(mat.dtype, np.integer):
+        mat[np.diag_indices(dim)] //= 2
+    else:
+        mat[np.diag_indices(dim)] /= 2
+
+    return mat
+
+
+def _sym_mat_to_vector(mat):
+    """Convert a symmetric matrix to a vector.
+
+    The upper triangle of the matrix (including the diagonal) will be used
+    as the values of the vector.
+
+    Parameters
+    ----------
+    mat : 2d-array
+        The symmetric matrix to convert to a vector
+
+    Returns
+    -------
+    vec : 1d-array
+        A vector consisting of the values of the upper triangle of the matrix.
+
+    See Also
+    --------
+    _vector_to_sym_mat
+    """
+    return mat[np.triu_indices_from(mat)]
+
+
+def read_csd(fname):
+    """Read a CrossSpectralDensity object from an HDF5 file.
+
+    Parameters
+    ----------
+    fname : str
+        The name of the file to read the CSD from. The extension '.h5' will be
+        appended if the given filename doesn't have it already.
+
+    Returns
+    -------
+    csd : instance of CrossSpectralDensity
+        The CSD that was stored in the file.
+
+    See Also
+    --------
+    CrossSpectralDensity.save : For saving CSD objects
+    """
+    if not fname.endswith('.h5'):
+        fname += '.h5'
+
+    csd_dict = read_hdf5(fname, title='conpy')
+    return CrossSpectralDensity(**csd_dict)
+
+
+def pick_channels_csd(csd, include=[], exclude=[]):
+    """Pick channels from covariance matrix.
+
+    Parameters
+    ----------
+    csd : instance of CrossSpectralDensity
+        The CSD object to select the channels from.
+    include : list of string
+        List of channels to include (if empty, include all available).
+    exclude : list of string
+        Channels to exclude (if empty, do not exclude any).
+
+    Returns
+    -------
+    res : instance of CrossSpectralDensity
+        Cross-spectral density restricted to selected channels.
+    """
+    sel = pick_channels(csd.names, include=include, exclude=exclude)
+    data = []
+    for vec in csd._data.T:
+        mat = _vector_to_sym_mat(vec)
+        mat = mat[sel, :][:, sel]
+        data.append(_sym_mat_to_vector(mat))
+    names = [csd.names[i] for i in sel]
+
+    return CrossSpectralDensity(
+        data=np.array(data).T,
+        names=names,
+        tmin=csd.tmin,
+        tmax=csd.tmax,
+        frequencies=csd.frequencies,
+        n_fft=csd.n_fft,
+    )
 
 
 @verbose
 def csd_epochs(epochs, mode='multitaper', fmin=0, fmax=np.inf,
-               fsum=True, tmin=None, tmax=None, n_fft=None,
+               frequencies=None, fsum=True, tmin=None, tmax=None, n_fft=None,
                mt_bandwidth=None, mt_adaptive=False, mt_low_bias=True,
-               projs=None, verbose=None):
+               cwt_n_cycles=7, decim=1, picks=None, projs=None, n_jobs=1,
+               verbose=None):
```

Comment:
this is not consistent with psd functions. To follow the psd naming we should have:

csd_morlet
csd_multitaper

this would allow you to avoid having many parameters ignored depending of value of mode

Reply:
what you can do is to make it a private function and just expose 2 public functions depending on the value of mode.

Reply:
I totally agree with this. However, I thought you wanted to break the existing API as little as possible, so I kept all the warts of the exiting CSD API. If I'm free to break this API, I can make it much smoother than it is now :)

Reply:
+1